src/gui/MainWindow.cpp

```diff
@@ -282,15 +284,44 @@ void MainWindow::initXournalWidget() {
     scrollHandling->init(this->xournal->getWidget(), layout);
 }
 
-void MainWindow::setTouchscreenScrollingForDeviceMapping() {
+void MainWindow::setGtkTouchscreenScrollingForDeviceMapping() {
     for (InputDevice const& inputDevice: DeviceListHelper::getDeviceList(this->getControl()->getSettings())) {
         InputDeviceClass deviceClass = InputEvents::translateDeviceType(inputDevice.getName(), inputDevice.getSource(),
                                                                         this->getControl()->getSettings());
         if (inputDevice.getSource() == GDK_SOURCE_TOUCHSCREEN && deviceClass != INPUT_DEVICE_TOUCHSCREEN) {
-            gtk_scrolled_window_set_kinetic_scrolling(GTK_SCROLLED_WINDOW(winXournal), false);
-            break;
+            setGtkTouchscreenScrollingEnabled(false);
+
+            return;
         }
     }
+
+    setGtkTouchscreenScrollingEnabled(true);
+}
+
+void MainWindow::setGtkTouchscreenScrollingEnabled(bool enabled) {
+    if (enabled == gtkTouchscreenScrollingEnabled || usingTouchWorkaround || winXournal == nullptr) {
+        return;
+    }
+
+    gtkTouchscreenScrollingEnabled = enabled;
+
+    // Run on the main thread:
+    // 0 = timeout (run as soon as possible)
+    gdk_threads_add_timeout(
+            0,
+            [](gpointer data) -> gboolean {
+                MainWindow* self = static_cast<MainWindow*>(data);
+
+                gtk_scrolled_window_set_kinetic_scrolling(GTK_SCROLLED_WINDOW(self->winXournal),
+                                                          self->gtkTouchscreenScrollingEnabled);
+
+                return FALSE;
```

Comment:
Macros are discouraged since c++11 and mostly obsolete since c++17. Please use the in build types.

```suggestion
                return false;
```